Allocate all memory up front

Istarnion opened the pull request:
Commits move gradually towards fewer memory allocations. The final one uses a slight hack to allocate both the `t_game` struct and the full capacity `t_snake` array in one malloc, which is not very pretty, but results in a single allocation, and ensures tight packing in memory for all game data.
I believe this closes #13

veloek commented on game.c:
```diff
@@ -112,83 +95,85 @@ static void handle_input()
-    while (tail->next != NULL)
-        tail = tail->next;
+    // Ensure we have capacity to grow
+    if(game->snake_capacity <= game->snake_length)
```
Will this ever happen when `snake_capacity` is initialized to fit the whole screen? Seems perhaps unnecessary...

veloek commented on game.c:
```diff
@@ -56,18 +49,8 @@ static void initialize()
-    free(snake);
-}
-
 static void destroy()
```
The `destroy` function can probably just be removed now.

veloek commented on game.c:
```diff
@@ -320,8 +305,15 @@ t_game* new_game(unsigned short rows, unsigned short cols)
-    game = (t_game*)malloc(sizeof(t_game));
-    assert(game);
+    int max_snake_length = (rows-2) * (cols-2);
+    unsigned char *game_memory =
```
Could perhaps leave a comment to explain why it's done like this instead of allocating memory for game and snake separately.

veloek reviewed (approved):
Awesome! I'm learning a lot from reading your code :smiley:

Istarnion commented:
Removed the redundant destroy function, removed the dynamic expansion code in `grow` (that was dangerous bug, even), and added some explaination on the memory allocation, as well as a debug line so one can check how much memory the game actually allocates.

veloek commented on game.c:
```diff
@@ -305,11 +291,24 @@ t_game* new_game(unsigned short rows, unsigned short cols)
+        max_snake_length * sizeof(t_snake);
+    unsigned char *game_memory = (unsigned char *)malloc(game_memory_size);
     assert(game_memory);
+    debug("Allocated %u kB of memory for the game.\n", game_memory_size);
```
Isn't `game_memory_size` in bytes rather than kilobytes? Pass `game_memory_size >> 10` perhaps?
